TEP-0046: Finally tasks execution post pipelinerun timeout (#326)
tektoncd/community: teps/0047-finallytask-execution-post-timeout.md

tags, and then generate with `hack/update-toc.sh`.
-->

<!-- toc -->
- [# TEP-0047: Finally tasks execution post pipelinerun timeout](#-tep-0047-finally-tasks-execution-post-pipelinerun-timeout)
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
- [Test Plan](#test-plan)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

<!--
This section is incredibly important for producing high quality user-focused
documentation such as release notes or a development roadmap.  It should be
possible to collect this information before implementation begins in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself.
A good summary is probably at least a paragraph in length.
Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.
[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

> This TEP adresses issue [`#2980`](https://github.com/tektoncd/pipeline/issues/2989). 

Reviewer: typo 🙃  2980 == 2989

Author: thanks

      taskSpec:
        steps:
          - name: hello
            image: ubuntu
            script: |
              echo "Hello World!"
              sleep 10
    finally:
    - name: task2
      params:
        - name: echoStatus
          value: "$(tasks.task1.status)"
      taskSpec:
        params:
          - name: echoStatus
        steps:
          - name: verify-status
            image: ubuntu
            script: |
              if [ $(params.echoStatus) == "Succeeded" ]
              then
                echo " Hello World echoed successfully"
              fi
```

The finally task runs after the task completion and both execute normally.

| NAME                                                           | TASK NAME        | STARTED        | DURATION   | STATUS                 |
|----------------------------------------------------------------|------------------|----------------|------------|------------------------|
| ∙ hello-world-pipeline-run-with-timeout-task1-tp9dq          | task1          | 17 seconds ago | 4 seconds  | Succeeded              |
| ∙ hello-world-pipeline-run-with-timeout-task2-zpzfm | task2 | 47 seconds ago | 30 seconds | Succeeded |

> |                                                                |                  |                |            |                        |

Reviewer: Not a blocker here, but this is weird:

* dag task `Task1` with `sleep 10` has Duration `4 seconds`
* final task `Task2` without any delay has Duration of `30 seconds`

Author: Thanks for reviewing. Will check this out

The pipelinerun timeout take precedence over the task timeout. After 10s the task fails... And the finally task does not get the chance to execute.

For this reason, it is currently not possible to rely on Finally tasks for any of the aforementioned use cases.

### Goals

<!--
List the specific goals of the TEP.  What is it trying to achieve?  How will we
know that this has succeeded?
-->

Enable the uses cases :

- Cleanup cluster resources after finishing (with success/failure) integration tests (Dogfooding Scenario)
- Update Pull Request with what happened overall in the pipeline (pipeline level)
- Report Test Results at the end of the test pipeline (Notifications Scenario)

When a pipelinerun times out.

### Non-Goals

<!--
What is out of scope for this TEP?  Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation.  The "Design Details" section below is for the real
nitty-gritty.
-->

> Enable finally task to run when a pipeline times out. This implies a behavioral change, as finally tasks will run no matter what. 

Reviewer 2: We will need to put this change of behaviour behind a feature flag, which will be default preserve the current behaviour. I think a possible solution could be as follows:
- when disabled (default), the new API field is not accepted via validation. The timeout is considered as an overall pipeline timeout. If it happens before finally, finally does not run, if it happens during finally, finally is interrupted and fails. 
- when enabled, the new API field is accepted via validation. The timeout is not applied to the pipeline excluding finally.

Alternatively, we could have two flags, one which controls the API change and one that controls how timeout is handled, to provide backward compatibility. In this scenario:
- if the API change is enabled (default to disabled), a finally timeout can be specified, but the overall behaviour depends by the other flag
- if the timeout backward compatibility flag is enabled (default to true), the timeout specified today applies to the overall pipeline. If a finally timeout is specified, the main pipeline will timeout after (overall timeout - finally timeout)

This second approach might make sense if we have a single flag to control all alpha API flag. One might want to enable alpha API parts but still maintain the current behaviour for pipeline timeouts.
@bobcatfish @vdemeester ^^^

Reviewer 3: Agreed, because this is a change in behaviour, we need to to preverve the current behaviour for a given amount of time.
What we are adding here is:
- A new field for timeout in `finally` task (with a default that could be the same default as for `TaskRun`)
- A change in behaviour so that the `pipelinerun` timeout doesn't apply to `finally` task.

I feel both can be seen as separate, as, we could have the 2nd without the first (just relying on the default taskrun timeout for `finally` task). At the same time, the 1st could be implemented independently of the behaviour change. Hence I am more on a flag on the controller to change the behavior, and add the timeout field under a general "alpha" feature flag — which is the 2nd approach @afrittoli.

Author: Not sure I understand here. Relying on the taskrun default timeout is in itself a behavioral change. So in based on the @afrittoli proposition 2 flags are still needed.

Reviewer 4: @souleb maybe the missing detail is that in [TEP-0033](https://github.com/tektoncd/community/pull/280#issuecomment-770923674) we are discussing adding an "alpha" feature flag that will need to be enabled to use all new alpha features, including specifying a timeout for finally tasks

i agree with @vdemeester and @afrittoli that it there are 2 separate features here and we can approach them differently

> Enable pipeline authors to specify a timeout field for finally tasks. In all normal run, that timeout is not needed and finally tasks execute after non-finally tasks. But in case of timed out pipeline, the finally task execution is bounded by the declared timeout.

Reviewer: > the finally task execution is bounded by the declared timeout

For a pipeline with a time out of 10 seconds, and finally tasks with a timeout 20 seconds (using a new field), pipeline stops running `dag` tasks after 10 seconds and starts running finally tasks. The finally tasks will run with 20 seconds of timeout after starting executing. In the end, pipeline overall will end up with a timeout of 30 seconds.

```yaml
spec:
  tasks:
    - name: tests
      taskRef:
        Name: integration-test
  finally:

>     timeout: "0h0m10s"

Reviewer: Please confirm, are you proposing a new field `timeout` under `finally` section which will apply to all finally tasks defined? Also what if each finally task has specified its own timeout plus this new field?

Author: yes exactly. It makes sense to me as all finally task are run in parallel. If a task has a timeout declared, the behavior is unchanged for the task, but the finally timeout field will take precedence.

If a task has a 20s timeout and the finally timeout field is 30s. Then the task timeout is relevant. The other way around, with a 30s task timeout and a 20s finally timeout, the task fails after 20s.

Reviewer 4: hmm this would be a backwards incompatible change, and would be very difficult as shown here. currently `finally` is a list, this example shows it as a hybrid of a dictionary and a list. if you wanted to have a `timeout` section in `finally` you'd probably need to add a tasks section or similar as well, e.g.:

```
finally:
  timeout: "0h0m10s"
  tasks:
  - ...
```

This would be a backwards incompatible change tho, so we'd need to support both for a while (@jerop was dealing with a similar issue for adding more configuration to when expressions - this feels like a lesson about the consequences of using lists).

Reviewer: yup this would be a backward incompatible change, instead how about adding a new field under `spec`:

```
kind: Pipeline
spec:
  finallyTimeout: ...
  tasks:
      ....
   finally:
      ...
```

or introduce a new field in `pipelineRun`.

Reviewer: we have two different options to specify timeout today (1) at the `pipelineTask` level (2) at the `pipelineRun` level. 

Timeout at the `pipelineRun` level:

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: hello-world-pipeline-run-with-timeout
spec:
  timeout: "0h0m60s"
```

Finally tasks are bound to this time out (or default `pipelineRun` timeout of 60 minutes). We are trying to override the `pipelineRun` `timeout` in this proposal. And, I recommend to add one more field at the same `pipelineRun` level which overrides `pipelineRun` timeout and the finally tasks would honor this new field:

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: hello-world-pipeline-run-with-timeout
spec:
  timeout: "0h0m60s"
  finallyTimeout: "0h0m60s"
```

Author: thanks @bobcatfish @pritidesai for the review.

I am wondering if adding a second timeout field at the `pipelinerun` spec level is the best strategy. As a user, it might be confusing, could make more difficult to understand the behavior.

What if we step back here and instead re-use the timeout field in the `pipelineTask`.  We would still override the `pipelinerun` timeout, but apply it for each finally task.

    - name: cleanup-test
      taskRef:
        Name: cleanup
```

> In order to keep older declaration working, we set the timeout field optional with a `default 20 minutes timeout`. 

Reviewer: why defaulting to 20 minutes?

Author: From the use cases, I would give the finally tasks a small time budget. I mean to are not intended for long batch operations. So 20m, but it's arbitrary.

  - [Finally block level timeout flag](#finally-block-level-timeout-flag)
  - [Pipelinerun timeout is inclusive of the finally tasks timeout](#pipelinerun-timeout-is-inclusive-of-the-finally-tasks-timeout)
  - [Finally Timeout flag at Pipelinerun Spec](#finally-timeout-flag-at-pipelinerun-spec)
<!-- /toc -->

## Summary

<!--
This section is incredibly important for producing high quality user-focused
documentation such as release notes or a development roadmap.  It should be
possible to collect this information before implementation begins in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself.
A good summary is probably at least a paragraph in length.
Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.
[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

This TEP adresses issue [`#2989`](https://github.com/tektoncd/pipeline/issues/2989). 

> The proposal is to enable finally tasks to execute when a pipelinerun has reached timeout.

Reviewer 5: ```suggestion
The proposal is to enable finally tasks to execute when the non-finally tasks have timed out.
```

## Motivation

<!--
This section is for explicitly listing the motivation, goals and non-goals of
this TEP.  Describe why the change is important and the benefits to users.  The
motivation section can optionally provide links to [experience reports][] to
demonstrate the interest in a TEP within the wider Tekton community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

The finally task [`design document`](https://docs.google.com/document/d/1lxpYQHppiWOxsn4arqbwAFDo4T0-LCqpNa6p-TJdHrw/edit#heading=h.w51ed6k2inef) list the following use cases :

- Cleanup cluster resources after finishing (with success/failure) integration tests (Dogfooding Scenario)
- Update Pull Request with what happened overall in the pipeline (pipeline level)
- Report Test Results at the end of the test pipeline (Notifications Scenario)

> Unfortunately if a pipeline's execution reaches the defined timeout value, the pipelinerun stop and reports a failed status.

Reviewer 5: ```suggestion
Unfortunately if a pipeline's execution reaches the defined timeout value when executing the non-finally tasks, the pipelinerun stop and reports a failed status without executing the finally tasks.
```

### Goals

<!--
List the specific goals of the TEP.  What is it trying to achieve?  How will we
know that this has succeeded?
-->

Enable the uses cases :

- Cleanup cluster resources after finishing (with success/failure) integration tests (Dogfooding Scenario)
- Update Pull Request with what happened overall in the pipeline (pipeline level)
- Report Test Results at the end of the test pipeline (Notifications Scenario)

When a pipelinerun times out.

### Non-Goals

<!--
What is out of scope for this TEP?  Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation.  The "Design Details" section below is for the real
nitty-gritty.
-->

Enable finally task to run when a pipeline times out.

We propose to deprecate the current `timeout` as the pipelinerun level. This timeout is as of today used as the timeout of each newly created taskrun by the controller.

Instead we propose to have to new flags: 
- tasksTimeout
- runTimeout

> `tasksTimeout` will define a timeout for the dag tasks. `runTimeout` will define a timeout for the whole pipelinerun. The finally tasks timeout willbe `runTimeout - tasksTimeout` with `runTimeout > tasksTimeout`.

Reviewer 2: NIT: willbe -> will be

When `tasksTimeout` is not defined, `runTimeout` is used for the tasks timeout.

This will enable users to manage run time behavior and make sure their finally tasks run as intended by scoping the tasks runtime period.

.
```yaml
spec:
  tasksTimeout: "1h0m0s"
  runTimeout: "2h0m0s"
  pipelineSpec:
    tasks:
    - name: tests
      taskRef:
        Name: integration-test
    finally:
    - name: cleanup-test
      taskRef:
        Name: cleanup
```

The default default-timeout-minutes configurable via configmap is kept and will be applied to the `runTimeout`.

In order to keep older declaration working. We will use the [`Feature Flag for all alpha features`](https://github.com/tektoncd/community/pull/280) to enable the new feature.

> When we opt-in the feature, we may specify a `tasksTimeout` and a `runTimeout` flags. The will be used to define dags tasks timeout and the pipelinerun timeout. We cannot use the current `timeout` flag. It would result in a validation error.

Reviewer 2: What would be the upgrade path with this approach if I have pipelines with `timeout` specified?

I think I would have to:
- delete `timeout` from existing pipelines
- enable the feature flag
- add `runTimeout` to pipelines

Possibly a different approach would be to deprecate "timeout" - if the feature flag is enabled it will be considered as a `runTimeout` instead, but if `tasksTimeout` or `runTimeout` are added, `timeout` has to be removed or it won't pass validation.

Author: We were thinking about the first approach. If the feature flag is enabled and you have a `timeout` specified, then you have a validation error.

So as you said, delete and replace it with `runTimeout`.

---
status: proposed
title: Finally tasks execution post pipelinerun timeout
creation-date: '2021-01-26'

> last-updated: '2021-01-26'

Reviewer: NIT: update the last-updated please

authors:
- '@souleb'
---

# TEP-0047: Finally tasks execution post pipelinerun timeout
---

<!-- toc -->

> - [# TEP-0047: Finally tasks execution post pipelinerun timeout](#-tep-0047-finally-tasks-execution-post-pipelinerun-timeout)

Reviewer: NIT: I don't think this is needed, we can delete this 🤔

Reviewer 5: you can use [`hack/update-toc.sh`](https://github.com/tektoncd/community/blob/main/hack/update-toc.sh) to autofix table of contents

| ∙ hello-world-pipeline-run-with-timeout-task1-q7fw4 | task1     | 2 minutes ago | 30 seconds | Failed(TaskRunTimeout) |
|                                                     |           |               |            |                        |
|                                                     |           |               |            |                        |

The pipelinerun timeout take precedence over the task timeout. After 10s the task fails... And the finally task does not get the chance to execute.

For this reason, it is currently not possible to rely on Finally tasks for any of the aforementioned use cases.

### Goals

<!--
List the specific goals of the TEP.  What is it trying to achieve?  How will we
know that this has succeeded?
-->

Enable the uses cases :

- Cleanup cluster resources after finishing (with success/failure) integration tests (Dogfooding Scenario)
- Update Pull Request with what happened overall in the pipeline (pipeline level)
- Report Test Results at the end of the test pipeline (Notifications Scenario)

When a pipelinerun times out.

> ### Non-Goals

Reviewer: NIT: this can be dropped

<!--
What is out of scope for this TEP?  Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation.  The "Design Details" section below is for the real
nitty-gritty.
-->

Enable finally task to run when a pipeline times out.

We propose to have a new flag ```timeouts``` which would be a dictionary of a set of timeouts. We propose adding ```tasks``` as the first key of that dictionary. 

> 

Reviewer: please clarify a new section being added to `pipelineRun` CRD, am I reading it right?

```yaml
timeouts:
  tasks: "0h4m0s"
  ...
```

> `tasks` will define a timeout for the dag tasks. The finally tasks timeout will be `timeout - tasks` with ```timeout```being the current timeout flag and `timeout >= tasks`.

Reviewer: I feel this is overkill, since we all agree to go in a direction with a new section as a dictionary, we can just isolate this without mixing with the existing `timeout` field in `pipelineRun`. Keep the existing behavior as is. Introduce a new section `timeouts` in `pipelineRun` to read `tasks` timeout, `pipeline` timeout, and `finally` timeout. Users can either specify the traditional timeout or this new timeout dictionary. 

```
kind: PipelineRun
spec:
  timeout: "0h4m0s"
```

or 

```
kind: PipelineRun
spec:
  timeouts:
    pipeline: "0h4m0s"
    tasks: "0h1m0s"
    finally: "0h3m0s"
```

Reviewer: My concern with mixing the new field and the traditional usage:

(1) Decoupling both in the future i.e replace `timeout` with `timeouts.pipeline`.
(2) Allowing to specify both `timeout` and `timeouts` right now and then restrict the same usage in the future.

Author: added a follow-on work section with the timeout dict proposition., Kept the proposal on the tasksTimeout field.
